src/ipc/handler.c

```diff
@@ -1250,31 +1228,43 @@ static inline struct ipc_msg *msg_find(struct ipc *ipc, uint32_t header,
 }
 
 int ipc_queue_host_message(struct ipc *ipc, uint32_t header, void *tx_data,
-			   size_t tx_bytes, uint32_t replace)
+			   size_t tx_bytes, bool replace)
 {
 	struct ipc_msg *msg = NULL;
-	uint32_t flags, found = 0;
+	bool found = false;
 	int ret = 0;
+	uint32_t flags;
 
 	ipc = cache_to_uncache(ipc);
 
 	spin_lock_irq(ipc->lock, flags);
 
 	/* do we need to replace an existing message? */
-	if (replace)
+	if (replace) {
 		msg = msg_find(ipc, header, tx_data);
+		found = (msg != NULL);
+	}
 
-	/* do we need to use a new empty message? */
-	if (msg)
-		found = 1;
-	else
-		msg = msg_get_empty(ipc);
+	if (!msg) {
+		/* max queue size exceeded */
+		if (ipc->shared_ctx->queue_size >= MSG_QUEUE_SIZE) {
+			trace_ipc_error(
+				"ipc_queue_host_message() error: queue size exceeded for msg hdr 0x%08x",
+				header);
+			ret = -EBUSY;
+			goto out;
+		}
 
-	if (msg == NULL) {
-		trace_ipc_error("ipc: msg hdr for 0x%08x not found "
-				"replace %d", header, replace);
-		ret = -EBUSY;
-		goto out;
+		/* create message */
+		msg = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
+			      SOF_MEM_CAPS_RAM, sizeof(*msg));
```

Comment:
@dbaluta I think allocating memory once per IPC may be a concern since that's quite a few allocations, right? What do you think?

Reply:
1. Stream positions are not sent every 1 ms, but rather every period specified by host (I've noticed that 40 ms is popular value).
2. Position updates are also written into memory window, so on the platforms with such capability we should really stop spamming IPCs.
3. The queue is limited to 12 messages anyway, so right now you also can lose the notification if the host won't process it fast enough or DSP has no idle cycles to go through queue. Nothing changes in this department and I would even go as far as saying that having full queue is more probable than not having free space on runtime heap.

Reply:
Yeah the issue is if there is some weird issue with the modules/components (future work) we may overallocate runtime memory and end up not having enough memory to be able to send 12 messages. I mean I'm not blocking over it but there are clients which may complain about this.

Allocating at prepare time may be acceptable, but playback stopping due to some leak from a component which isn't even upstreamed means reliability takes a (small) dip.

Your point about the memory window, yeah I think I should figure out how to make a memory window myself. Is there some section of the mailbox itself that I can reuse or I should update the memory map (taking advantage of the huge shared memory on i.MX) and just add a new region specifically for this? How exactly does it work?

"you can also lose the notification if [...] DSP has no idle cycles to go through queue" That is easier to hit during testing though -- in fact hitting it made me enable the caches when it had to happen. Still, losing the notification will cause serious problems of quality. The allocation issues may in certain situations which third party code be hit only under certain circumstances.

Reply:
@tlauda @paulstelian97 I think Paul has a point in this but because the queue size is pretty small I don't think it matters. If allocations will start failing at runtime we are in trouble. Another, point to consider is that allocations will take time. 

@paulstelian97 I do think it will worth considering by using the shared memory to post position updates instead of sending them through IPCs.

Reply:
@paulstelian97 @dbaluta Changed it back to for now. We can revisit this in the future.